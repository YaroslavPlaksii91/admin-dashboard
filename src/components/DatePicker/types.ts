import { FieldErrors, FieldValues } from 'react-hook-form';

export type DatePickerProps = {
  errors: FieldErrors;
  field: FieldValues;
  label: string;
  name: string;
};
