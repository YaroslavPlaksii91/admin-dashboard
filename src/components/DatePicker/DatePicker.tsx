import { FC } from 'react';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { DateTimePicker, LocalizationProvider } from '@mui/x-date-pickers';

import { DatePickerProps } from './types';

export const DatePicker: FC<DatePickerProps> = ({
  errors,
  field,
  label,
  name,
  disabled,
}) => {
  const errorMessage = errors[name]?.message?.toString();

  return (
    <LocalizationProvider dateAdapter={AdapterDayjs}>
      <DateTimePicker
        label={label}
        value={field.value || null}
        inputRef={field.ref}
        onChange={field.onChange}
        onAccept={field.onChange}
        disabled={disabled}
        slotProps={{
          textField: {
            variant: 'outlined',
            fullWidth: true,
            InputLabelProps: { shrink: true },
            error: !!errors[name],
            helperText: errorMessage || '',
          },
        }}
      />
    </LocalizationProvider>
  );
};
