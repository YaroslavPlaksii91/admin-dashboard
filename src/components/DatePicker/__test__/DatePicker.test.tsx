import { cleanup, render, screen } from '@testing-library/react';
import { FieldErrors, FieldError } from 'react-hook-form';

import { DatePicker } from '../DatePicker';

describe('DatePicker component', () => {
  afterEach(cleanup);

  const label = 'Label';
  const name = 'datePicker';
  const errors = {};
  const field = {};
  const isDisabled = false;

  it('should be rendered', () => {
    const { asFragment } = render(
      <DatePicker
        label={label}
        name={name}
        errors={errors}
        field={field}
        disabled={isDisabled}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(
      <DatePicker
        label={label}
        name={name}
        errors={errors}
        field={field}
        disabled={isDisabled}
      />,
    );

    const datePickerLabel = screen.getByLabelText(label);
    const datePickerInput = screen.getByRole('textbox', { name: label });

    expect(datePickerLabel).toBeInTheDocument();
    expect(datePickerInput).toBeInTheDocument();
  });

  it('displays error message when there are errors', () => {
    const errorMessage = 'Invalid date';
    const errors: FieldErrors = {
      [name]: { message: errorMessage } as FieldError,
    };

    render(
      <DatePicker
        label={label}
        name={name}
        errors={errors}
        field={field}
        disabled={isDisabled}
      />,
    );

    const errorText = screen.getByText(errorMessage);

    expect(errorText).toBeInTheDocument();
  });

  it('disables the date picker when disabled prop is true', () => {
    const disabled = true;

    render(
      <DatePicker
        label={label}
        name={name}
        errors={errors}
        field={field}
        disabled={disabled}
      />,
    );

    const datePickerInput = screen.getByRole('textbox', { name: label });

    expect(datePickerInput).toBeDisabled();
  });
});
