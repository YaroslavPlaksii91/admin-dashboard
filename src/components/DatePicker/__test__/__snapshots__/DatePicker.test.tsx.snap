// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Label
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        inputmode="text"
        placeholder="MM/DD/YYYY hh:mm aa"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Choose date"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
